File: cracking-the-coding-interview/testcase.js
var fs = require('fs');
var path = require('path');

var tc = {
    list: function(folder){
        var files = fs.readdirSync(folder);
        var inputs = files.filter(filename => filename.startsWith("input"));
        var outputs = files.filter(filename => filename.startsWith("output"));
        var testcases = inputs.map(function (item, i) { 
            return {
                input: path.join(folder, item), 
                output: path.join(folder, outputs[i])
            } 
        });
        return testcases;
    },
    run: function(testcase, runFn){
        var input = fs.readFileSync(testcase.input, "utf8");
        var begin = Date.now();
        var result = runFn(input.split("\n"));
        var end = Date.now();
        var expected = fs.readFileSync(testcase.output, "utf8");
        return {
             success: (expected === result),
             runtime: (end-begin) + " msec",
              result: result,
            expected: expected
        }
    },
    suite: function(testcases, fn){
        var that = this;
        return testcases.map(
            function(testcase, i) {
                var tc = that.run(testcase, fn);
                return {
                    testcase: "#" + i,
                     success: tc.success,
                     runtime: tc.runtime
                };
            });
    },
    results: function(tests){
        var titles = ["Test case", "Success", "Runtime"];
        var results = tests.map(
            function(test){
                return [test.testcase, test.success, test.runtime]
            }
        );
        results.unshift(titles);
        return results;
    }
}

module.exports = tc

File: cracking-the-coding-interview/testcase.py
import os
from timeit import default_timer as timer
from math import ceil

def merge(d1, d2):
    ''' Merge two dictionaries. '''
    merged = {}
    merged.update(d1)
    merged.update(d2)
    return merged

def filenames(prefix, folder):
    return [{prefix: os.path.join(folder, file)} 
        for file in os.listdir(folder) 
            if file.startswith(prefix)]

def elapsed_str(end, start):
    return str(int(round((end-start)*1000)))+" msec"

class TestCase:
    def __init__(self, folder):
        self.folder = folder

    def list(self):
        inputs  = filenames("input", self.folder)
        outputs = filenames("output", self.folder)
        return [merge(i,o) for i,o in list(zip(inputs, outputs))]

    def run(self, testcase, fn):
        with open(testcase["input"]) as inf, open(testcase["output"]) as outf:
            input_text = inf.read()
            start = timer()
            result = " ".join(str(x) for x in fn(input_text.split("\n")))
            end = timer()
            expected_text = outf.read()
            return {
                "success": (result == expected_text),
                "runtime": elapsed_str(end, start),
                "result": result,
                "expected": expected_text
            }

    def suite(self, testcases, fn):
        return [self.fields(self.run(test, fn)) for test in testcases]

    def results(self, tests):
        titles = [["Test case", "Success", "Runtime"]]
        results = [["#%s" % i, t["success"], t["runtime"]] for i, t in enumerate(tests)]
        return titles + results

    def fields(self, item):
        return {
            "success": item["success"],
            "runtime": item["runtime"]
        }
